libs/langchain/langchain/embeddings/vertexai.py

```diff
@@ -1,46 +1,298 @@
-from typing import Dict, List
+import logging
+import re
+import string
+import threading
+from concurrent.futures import ThreadPoolExecutor, wait
+from typing import Any, Dict, List, Literal, Optional, Tuple
 
 from langchain_core.embeddings import Embeddings
 from langchain_core.pydantic_v1 import root_validator
 
+from langchain.llms.base import create_base_retry_decorator
 from langchain.llms.vertexai import _VertexAICommon
 from langchain.utilities.vertexai import raise_vertex_import_error
 
+logger = logging.getLogger(__name__)
+
+_MAX_TOKENS_PER_BATCH = 20000
+_MAX_BATCH_SIZE = 250
+_MIN_BATCH_SIZE = 5
+
 
 class VertexAIEmbeddings(_VertexAICommon, Embeddings):
     """Google Cloud VertexAI embedding models."""
 
     model_name: str = "textembedding-gecko"
 
+    # https://cloud.google.com/vertex-ai/docs/generative-ai/embeddings/get-text-embeddings#api_changes_to_models_released_on_or_after_august_2023
+    embeddings_type: Optional[
+        Literal[
+            "RETRIEVAL_QUERY",
+            "RETRIEVAL_DOCUMENT",
+            "SEMANTIC_SIMILARITY",
+            "CLASSIFICATION",
+            "CLUSTERING",
+        ]
+    ] = None
+
+    # instance context
+    instance: Dict[str, Any] = {}  #: :meta private:
+
     @root_validator()
     def validate_environment(cls, values: Dict) -> Dict:
         """Validates that the python package exists in environment."""
         cls._try_init_vertexai(values)
         try:
             from vertexai.language_models import TextEmbeddingModel
+
+            values["client"] = TextEmbeddingModel.from_pretrained(values["model_name"])
         except ImportError:
             raise_vertex_import_error()
-        values["client"] = TextEmbeddingModel.from_pretrained(values["model_name"])
         return values
 
+    def __init__(
+        self,
+        project: Optional[str] = None,
+        location: str = "us-central1",
+        request_parallelism: int = 5,
+        max_retries: int = 6,
+        model_name: str = "textembedding-gecko",
+        credentials: Optional[Any] = None,
+        embeddings_type: Optional[
```

Comment:
seems unintuitive that embed_documents would use anything other than RETRIEVAL_DOCUMENT and embed_query would use anything other than RETRIEVAL_QUERY. should those use hard coded embeddings_type, and then if we want could expose a generic `embed` method that uses configured embeddings_type?

Reply:
@baskaryan I made a slight change to accommodate that.
`VertexAIEmbeddings` doesn't carry that parameter anymore. Instead, it only detects whether the embedding task type can be used - based on the endpoint name.

`embed_query` and `embed_documents` use the most suitable embeddings type for their respective task when calling a new `embed` method. 
`embed` accepts an embeddings task type name as a parameter. In all cases, the parameter is only passed to the model when it's allowed, or silently ignored otherwise.

Please take a look.

Reply:
@baskaryan Please let me know if there is anything that should be improved here. Thank you!

Reply:
@baskaryan I moved files to the new `libs/community`. Please take a look.